products/batch_exports/backend/temporal/pipeline/transformer.py

```diff
@@ -606,13 +609,21 @@ def remove_escaped_whitespace_recursive(value):
             return value
 
 
-def ensure_curly_brackets_array(v: list[typing.Any]) -> str:
-    """Convert list to str and replace ends with curly braces for PostgreSQL arrays.
+def _ensure_curly_brackets_array(v: list[typing.Any]) -> str:
+    """Convert list to PostgreSQL array literal format with proper escaping.
 
-    NOTE: This doesn't support nested arrays (i.e. multi-dimensional arrays).
+    Supports nested arrays and properly escapes special characters.
+    Uses psycopg3's ListDumper for correct PostgreSQL array formatting.
     """
-    str_list = str(v)
-    return f"{{{str_list[1:len(str_list)-1]}}}"
+
+    tx = psycopg.adapt.Transformer()
+    dumper = psycopg.types.array.ListDumper(list, tx)
+    result = dumper.dump(v)
+    assert result is not None
```

Comment:
It's a bit annoying that we need this assertion. [Looking at the code](https://github.com/psycopg/psycopg/blob/master/psycopg/psycopg/types/array.py#L157C39-L157C45), I can't see any scenario in which `None` is returned: `dump` boils down to `return b"".join(tokens)`, which does not return `None`: https://docs.python.org/3/library/stdtypes.html#bytes.join.

Reply:
yeah, I think it's because they've [annotated the function](https://github.com/psycopg/psycopg/blob/master/psycopg/psycopg/types/array.py#L157) as possibly returning `None` for some reason 😞